tests/test_fit_result.py

```diff
@@ -37,81 +85,96 @@ def ge_constraint(a):  # Has to be in the global namespace for pickle.
     """
-    Tests the FitResults from fitting a simple model with Le as constraint
+    Tests the FitResults from fitting a simple model with Le as symbolic constraint
     using several different minimizers and objectives.
     """
 
     model=Model({y: a * x ** b})
-    constraints = [
-        Le(a, b)
-    ]
     fit = Fit(**fit_kwargs, minimizer=minimizer, model=model, constraints=constraints)
     fit_result = fit.execute()
-    _run_tests(fit_result, expected_par, expected_std, expected_obj, expected_got)
+    _run_tests(fit_result, expected_par, expected_std, expected_obj, expected_gof)
 
 
 @pytest.mark.parametrize('minimizer',
     # Added None to check the defaul objective
-    # Removed TrustConstr, because it cannot handle CNM
+    # Removed TrustConstr, because it cannot handle CallableNumericalModel because of missing hassian
     subclasses(ConstrainedMinimizer) - {TrustConstr} | {None}
 )
-@pytest.mark.parametrize('fit_kwargs, expected_par, expected_std, expected_obj, expected_got',
+@pytest.mark.parametrize('fit_kwargs, expected_par, expected_std, expected_obj, expected_gof',
     [
         # No special objective
-        (dict(x=np.linspace(1, 10, 10), y=3 * np.linspace(1, 10, 10) ** 2), {a: 2.152889, b: 2.152889}, 
-         {a: 0.23715, b: 0.05076}, LeastSquares, {'r_squared': 0.99791, 'objective_value': 98.83587, 'chi_squared': 197.671746}),
+        (dict(x=np.linspace(1, 10, 10), y=3 * np.linspace(1, 10, 10) ** 2),
+         {a: 2.152889, b: 2.152889}, {a: 0.23715, b: 0.05076},
+         LeastSquares,
+         {'r_squared': 0.99791, 'objective_value': 98.83587, 'chi_squared': 197.671746}),
         # Test objective LeastSqueares
-        (dict(objective=LeastSquares, x=np.linspace(1, 10, 10), y=3 * np.linspace(1, 10, 10) ** 2), {a: 2.152889, b: 2.152889}, 
-         {a: 0.23715, b: 0.05076}, LeastSquares, {'r_squared': 0.99791, 'objective_value': 98.83587, 'chi_squared': 197.671746}),
+        (dict(objective=LeastSquares, x=np.linspace(1, 10, 10), y=3 * np.linspace(1, 10, 10) ** 2),
+         {a: 2.152889, b: 2.152889}, {a: 0.23715, b: 0.05076},
+         LeastSquares,
+         {'r_squared': 0.99791, 'objective_value': 98.83587, 'chi_squared': 197.671746}),
         # Test objective LogLikelihood
-        (dict(objective=LogLikelihood, x=np.linspace(1, 10, 10)), {a: 653.48460, b: 653.48460}, 
-         {a: float('nan'), b: float('nan')}, LogLikelihood, 
+        (dict(objective=LogLikelihood, x=np.linspace(1, 10, 10)),
+         {a: 653.48460, b: 653.48460}, {a: float('nan'), b: float('nan')},
+         LogLikelihood, 
          {'objective_value': -float('inf'), 'log_likelihood': float('inf'), 'likelihood': float('inf')})
     ])
-def test_constraints_CNM(minimizer, fit_kwargs, expected_par, expected_std, expected_obj, expected_got):
+def test_constraints_CNM(minimizer, fit_kwargs, expected_par, expected_std, expected_obj, expected_gof):
```

Comment:
Add the CNM constraint to the parametrization. To keep it from getting too unwieldy, do e.g.:
```python
CNM_CONSTRAINT = CallableNumericalModel.as_constraint(
    {z: ge_constraint}, connectivity_mapping={z: {a}},
    constraint_type=Ge, model=model
)

@pytest.mark.parametrize('constraints, ...', [
    ([Le(a, b), CNM_CONSTRAINT], ... ),
    ...
])
```

But that makes it look *a lot* like the other constraint tests, so it may be possible to deduplicate some code by calling `pytest.skip`:
```py
def test_constraints(...):
    ...
    if isinstance(minimizer, TrustConstraint) and CNM_CONSTRAINT in constraints:  # Maybe isinstance to check constraint classes as well?
        pytest.skip('TrustConstraint minimizer cannot handle CallableNumericalModel constraints since those have no Hessian')
    ...
```
Or call `pytest.xfail`, up to you.

Reply:
So you propose to merge `test_constraints` and `test_constraints_CNM`? :)

Reply:
In this case, I would add a new level in the parametrization-stack for constraints, because the results with the constraints `[Le(a,b)]` and `[Le(a,b), CNM_CONSTRAINT]` are equal (stdev, etc. too?).
The problem is, that we may have to rearrange the parametrization if we want to add new constraints with different results in the future. What do you mean?

Reply:
> In this case, I would add a new level in the parametrization-stack for constraints, because the results with the constraints [Le(a,b)] and [Le(a,b), CNM_CONSTRAINT] are equal (stdev, etc. too?).

Sure. Also add [CNM_CONSTRAINT] then.

> The problem is, that we may have to rearrange the parametrization if we want to add new constraints with different results in the future. What do you mean?

Let's leave tomorrow's problems for tomorrow to solve. Maybe pytest will get expanded with, I don't know, `example_parameters` where you specify all parameters again. Or whatever. For now just do what's easiest now.

> (btw. which constraint classes cannot be handled by TrustConstr?)

It's been too long since I've looked at the symfit code to give you an authoritative answer on this. You could also do a `hasattr(constraint, 'hessian')` as alternative maybe.

Reply:
> Also add [CNM_CONSTRAINT] then.

Then we have different results and this trick doesn't work anymore :/ 

> You could also do a hasattr(constraint, 'hessian') as alternative maybe.

Doesn't work :( Maybe we should leave it with `CNM_CONSTRAINT in constraints` then.

Reply:
> Then we have different results and this trick doesn't work anymore :/

Then splitting it out into different entries is fine with me.

> Doesn't work :( Maybe we should leave it with CNM_CONSTRAINT in constraints then.

`constraints` is a list, so it does require a bit more fiddling: `if ... or any(not hasattr(c, 'hessian') for c in constraints): ...`

Reply:
> constraints is a list, so it does require a bit more fiddling: if `... or any(not hasattr(c, 'hessian') for c in constraints)`: 

I tried it ;) CNM_CONSTRAINT  has only the attributes `['_connectivity_mapping', 'model_dict', 'dependent_vars', 'interdependent_vars', 'independent_vars', '_params', '__signature__', 'sigmas', 'constraint_type']`

> Then splitting it out into different entries is fine with me.

Do mean two test-methods? One with `[Le(a,b)], [Le(a,b), CNM_CONSTRAINT]`, the other with `[CNM_CONSTRAINT]` as  constraints (parametrized)?

Reply:
> Exactly, no 'hessian' attribute, so skip the TrustConstr test. Or mark it as xfail.

Oh. I thought it was the other way around :D sorry

> No, still 1 test method. E.g.

Got it ;) Thank you!

Reply:
> Exactly, no 'hessian' attribute, so skip the TrustConstr test. Or mark it as xfail.

Sorry for asking again, but `Le(a,b)` has no hessian attribute either :/
And the results of `COBYLA` with the constraint `[CNM_CONSTRAINT]` are about 0.5-1.0 off the expected value

Reply:
Hmmn. Ah. Oh. Hmmn. I guess that gets turned into a symbolic model with hessian further down, probably by Fit. `Le` should be an instance of a sympy Expression, right? So we could test for that.

Reply:
If there isn't a possibility that TrustConstr can't handle the constraint which is an instance of a sympy Expression (and the other way around: the constraint isn't an instance of a sympy Expression but TrustConstr can handle it) 🙆‍♂️ 
Otherwise we could test if any constraint equals `CNM_CONSTRAINT`...simple but functioning

Reply:
If and only if the constraint is a sympy expression we can take it's derivatives, which makes TrustConstr happy. In this case checking whether any of the contraints equals `CNM_CONSTRAINT` is a fine solution.

Reply:
> checking whether any of the contraints equals CNM_CONSTRAINT is a fine solution.

I'll add this one, because by checking whether the constraint is a sympy expression the readability decreases (checking if the constraint is an instance of Rational or Basic is kind of random in my opinion).

Reply:
> And the results of COBYLA with the constraint [CNM_CONSTRAINT] are about 0.5-1.0 off the expected value.

Concerning COBYLA: If I use `[CNM_CONSTRAINT]` as a constraint with COBYLA as minimizer, I got pretty wrong results. I expect `a` to be 3 and `b` to be 2, but I got the following results:
```
a:  2.3525773072081213
b:  2.111988688141647
a stdev:  0.190113214366974
b stdev:  0.03730180035197476
r_squared:  0.9988580222167673
```
new issue? :D

Reply:
Maybe. Isn't it just a non-convergence issue? i.e. the minimizer runs into it's maximum allowed number of evaluations and quits. It may be fixable by setting max_iter=int(1e6) and tol=1e-8. See also my comment on #324

Reply:
>  It may be fixable by setting max_iter=int(1e6) and tol=1e-8

Both issues (wrong results with and without constraints) can be fixed with more iterations. I'll think about a smart possibility to add `{options={'maxiter': int(1e8)}` to `fit.execute()` the next few days :D

Reply:
Excellent. Maybe add a `kwargs` to the parametrization, which can be empty for most cases?

Reply:
> for most cases

... would mean *for COBYLA as minimizer*. The easiest way would be an `if COBYLA`-statement in the test-method.
Adding `kwargs` to the minimizer-parametrization is difficult to read (we discussed this earlier regarding [minimizer and constraints](https://github.com/tBuLi/symfit/pull/289#issuecomment-792897758).

Another idea is a replacement of the minimizer-parametrization with a fixture with parameters. eg:
```python
@pytest.fixture(params=list(subclasses(BaseMinimizer)-....))
def minimizer(request):
    minimizer = request.param
    execute_kwargs = {}
    if minimizer is COBYLA:
        execute_kwargs['options'] = {'maxiter': int(1e8)}
    return (minimizer, execute_kwargs)
``` 

I have to admit that the test method itself could be a bit crazy at the beginning, but we could scale this pattern in the future :)

Reply:
Sure. I have no strong preference for one or the other. Experiment and pick your favourite I'd say.